### ISSUE TITLE
Using if foo -> \bar { } has severe performance penalties

### ISSUE THREAD

**Issue by lizmat** [rakudo/rakudo#3546]
$ raku -e 'my $a = 42; for ^1_000_000 { if $a -> \b { } }; say now - INIT now'
    0.20067557

    $ raku -e 'my $a = 42; for ^1_000_000 { if my \b = $a { } }; say now - INIT now'
    0.0196358

Aka, the Perl style of aliasing a result in an `if` is 10x faster.  :-(

Running `my $a = 42; for ^1_000_000 { if my \b = $a { } }` through the profiler, shows **NO** garbage collections, shows **71** garbage collections, most notable an `Int` and a `BootCode` allocation per iteration.

**Comment by lizmat** [comment]
FWIW, the profile states that the allocations happen in the `raku -e`, so that's not giving a lot of info to go on.

**Comment by lizmat** [comment]
Hmmm... I was wondering whether this performance difference was caused by generic `-> { }` overhead?  And that `if { }` is basically highly optimized?

**Comment by jnthn** [comment]
> FWIW, the profile states that the allocations happen in the raku -e, so that's not giving a lot of info to go on.

The allocations will be closure clones; I can't really think of a more sensible place for those to show up than in the enclosing block, which is what causes the clone to happen. So I'd say it's accurate.

> Hmmm... I was wondering whether this performance difference was caused by generic -> { } overhead? And that if { } is basically highly optimized?

This can be explored by turning off optimization. If we run the two with static optimization disabled:

```
$ perl6 --optimize=off -e 'my $a = "foo"; for ^1_000_000 { if my \b = $a { } }; say now - INIT now'
0.519067024
$ perl6 --optimize=off -e 'my $a = "foo"; for ^1_000_000 { if $a -> \b { } }; say now - INIT now'
0.5508771
```

Then we observe they are very close indeed. For comparison, here are the optimized times:

```
$ perl6 -e 'my $a = "foo"; for ^1_000_000 { if $a -> \b { } }; say now - INIT now'
0.33430947
$ perl6 -e 'my $a = "foo"; for ^1_000_000 { if my \b = $a { } }; say now - INIT now'
0.0428758
```

Since I worked on the optimizer, I have a good guess at what's going on: scope flattening. When it thinks it can get away with it, the optimizer does away with the inner scope. We've gradually made this more aggressive over time, because as can clearly be seen here it's quite worthwhile (even if this is a rather artificial benchmark - you'd do something in the block normally). However, so far it never attempts that on a block with parameters (aside: I'm pretty sure the NQP one already does know how to do deal with this, but NQP is rather simpler). Since the block is empty - which it'd not be in a real program - we're basically measuring the cost of taking a million closures and GC'ing them later.

**Comment by MasterDuke17** [comment]
Heh, for the heck of it I tried commenting out the `&& $block.arity == 0` here https://github.com/MasterDuke17/rakudo/blob/master/src/Perl6/Optimizer.nqp#L1158. Rakudo built ok and passed `make m-test m-spectest`, but it didn't make a difference for the example code.

**Comment by jnthn** [comment]
@MasterDuke17 That check is not really the important one; it's a heuristic to stop us wasting more time on a situation that's only going to be hopeless anyway. It's the "does it take a parameter" check that really rules it out:

https://github.com/MasterDuke17/rakudo/blob/master/src/Perl6/Optimizer.nqp#L1175

**Comment by jnthn** [comment]
Also, I misremembered-ish: NQP doesn't really do much about this. It only has a special case for `for` loops.

**Comment by MasterDuke17** [comment]
Yeah, without really knowing what I was doing I tried adding an additional
condition that `$_.name` must start with '__lowered_lex_' but that didn't
work (neither did the opposite of must start with '__lowered_param_').
